replicator/active_replicator.go

```diff
@@ -0,0 +1,134 @@
+	Connect() error
+	// Start will trigger the replicator to start its replication for the connection.
+	Start() error
+	// Close stops and closes any ongoing replications and connections.
+	Close() error
+}
+
+// BidirectionalActiveReplicator is a wrapper to encapsulate separate push and pull active replicators.
+type BidirectionalActiveReplicator struct {
+	// push *ActivePushReplicator // TODO: CBG-784
+	pull *ActivePullReplicator
+}
+
+// Compile-time interface check.
+var _ ActiveReplicator = &BidirectionalActiveReplicator{}
+
+// NewBidirectionalActiveReplicator returns a bidirectional active replicator for the given config.
+func NewBidirectionalActiveReplicator(ctx context.Context, config *ActiveReplicatorConfig) (*BidirectionalActiveReplicator, error) {
+	if errs := config.Validate(); len(errs) > 0 {
+		return nil, fmt.Errorf("%v", errs)
+	}
+
+	bar := &BidirectionalActiveReplicator{}
+
+	// if pushReplication := config.Direction == ActiveReplicatorTypePush || config.Direction == ActiveReplicatorTypePushAndPull; pushReplication {
+	// 	bar.Push = NewPushReplicator(config)
+	// }
+
+	if pullReplication := config.Direction == ActiveReplicatorTypePull || config.Direction == ActiveReplicatorTypePushAndPull; pullReplication {
+		bar.pull = NewPullReplicator(config)
+	}
+
+	return bar, nil
+}
+
+func (bar *BidirectionalActiveReplicator) Connect() error {
+	// if bar.push != nil {
+	// 	if err := bar.push.Connect(); err != nil {
+	// 		return err
+	// 	}
+	// }
+
+	if bar.pull != nil {
+		if err := bar.pull.Connect(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (bar *BidirectionalActiveReplicator) Start() error {
+	// if bar.push != nil {
+	// 	if err := bar.push.Start(); err != nil {
+	// 		return err
+	// 	}
+	// }
+
+	if bar.pull != nil {
+		if err := bar.pull.Start(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (bar *BidirectionalActiveReplicator) Close() error {
+	// if bar.push != nil {
+	// 	if err := bar.push.Close(); err != nil {
+	// 		return err
+	// 	}
+	// }
+
+	if bar.pull != nil {
+		if err := bar.pull.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// blipSync opens a connection to the target, and returns a blip.Sender to send messages over.
+func blipSync(target url.URL, blipContext *blip.Context) (*blip.Sender, error) {
+	// GET target database endpoint to see if reachable for exit-early/clearer error message
+	resp, err := http.Get(target.String())
+	if err != nil {
+		return nil, err
+	}
+
+	err = resp.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code %d from target database", resp.StatusCode)
+	}
+
+	// switch to websocket protocol scheme
+	if target.Scheme == "http" {
+		target.Scheme = "ws"
+	} else if target.Scheme == "https" {
+		target.Scheme = "wss"
+	}
+
+	config, err := websocket.NewConfig(target.String()+"/_blipsync", "http://localhost")
```

Comment:
Spent a bit of time investigating when writing this, and we have a couple of options:

- Continue as we do today, and ignore Origin on the passive side.
- Implement a config option for allowed origins on the passive side for SG-SG replications.
We can't really add this for CBL connections, because it's usually impossible to know what origins your clients are connecting from ahead of time.
Seems like this client validation for SG-to-SG replications would be better served by x509 auth.

For the first option, we still have to set a valid URL. It's non-trivial to get an accurate origin address without using something like `base.FindPrimaryAddr()`, which doesn't seem worth it if it's not being used on the passive side.